ocaml/typing/typecore.ml

```diff
@@ -3767,6 +3778,14 @@ and type_approx_aux_jane_syntax : Jane_syntax.Expression.t -> _ = function
   | Jexp_immutable_array _
   | Jexp_unboxed_constant _ -> ()
 
+and type_approx_aux_case_rhs env rhs in_function ty_expected =
+  match rhs with
+  | Psimple_rhs e | Pboolean_guarded_rhs { pbg_rhs = e; _ } ->
+      type_approx_aux env e in_function ty_expected
+  | Ppattern_guarded_rhs { ppg_scrutinee; ppg_cases; ppg_loc } ->
+      let match_desc = Pexp_match (ppg_scrutinee, ppg_cases) in
+      type_approx_aux_desc env match_desc ppg_loc in_function ty_expected
```

Comment:
I think this is an improvement over the previous code, but I'd still like to avoid constructing a new `Pexp_match` node -- can we instead inline the bit of code from `type_approx_aux_desc` (or factor it out to a helper)?

Reply:
I inlined, which seemed appealing because the logic for Pexp_match is pretty simple.